{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "sys.path.append('C:\\\\Users\\\\Dell\\\\Desktop\\\\CV Projects\\\\prostate_cancer_genomics')\n",
    "\n",
    "from config import PROSTATE_DATA_PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = os.path.join(PROSTATE_DATA_PATH, 'raw_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = '41588_2018_78_MOESM4_ESM.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>type</th>\n",
       "      <th>classification</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>patient</th>\n",
       "      <th>Primary_Met</th>\n",
       "      <th>pair_id</th>\n",
       "      <th>individual_id</th>\n",
       "      <th>case_sample</th>\n",
       "      <th>Pair_Set_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>58862934</td>\n",
       "      <td>58862934</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGA-EJ-5499</td>\n",
       "      <td>PRAD-TCGA-EJ-5499-Normal-SM-1U3NZ</td>\n",
       "      <td>...</td>\n",
       "      <td>58862934</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>PRAD-TCGA-EJ-5499-Tumor-SM-1U3IG</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PRAD-TCGA-EJ-5499-TP-NB-SM-1U3IG-SM-1U3NZ</td>\n",
       "      <td>PRAD-TCGA-EJ-5499</td>\n",
       "      <td>PRAD-TCGA-EJ-5499-Tumor-SM-1U3IG</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>58863660</td>\n",
       "      <td>58863660</td>\n",
       "      <td>+</td>\n",
       "      <td>MO_1012</td>\n",
       "      <td>MO_1012-Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>58863660</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>MO_1012-Tumor-Abdomen_wall_nodule</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>MO_1012_TM_NB_MO_1012-Tumor-Abdomen_wall_nodul...</td>\n",
       "      <td>MO_1012</td>\n",
       "      <td>MO_1012-Tumor-Abdomen_wall_nodule</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>58863782</td>\n",
       "      <td>58863782</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGA-CH-5752</td>\n",
       "      <td>PRAD-TCGA-CH-5752-Normal-SM-1U3IX</td>\n",
       "      <td>...</td>\n",
       "      <td>58863782</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>PRAD-TCGA-CH-5752-Tumor-SM-1U3ID</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PRAD-TCGA-CH-5752-TP-NB-SM-1U3ID-SM-1U3IX</td>\n",
       "      <td>PRAD-TCGA-CH-5752</td>\n",
       "      <td>PRAD-TCGA-CH-5752-Tumor-SM-1U3ID</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>58864304</td>\n",
       "      <td>58864304</td>\n",
       "      <td>+</td>\n",
       "      <td>06-134H1_LN</td>\n",
       "      <td>06-134A1_NORMAL</td>\n",
       "      <td>...</td>\n",
       "      <td>58864304</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>06-134H1_LN</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>06-134H1_LN_06-134A1_NORMAL</td>\n",
       "      <td>06-134</td>\n",
       "      <td>06-134H1_LN</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>10</td>\n",
       "      <td>52569681</td>\n",
       "      <td>52569681</td>\n",
       "      <td>+</td>\n",
       "      <td>SC_9126</td>\n",
       "      <td>SC_9126_Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>52569681</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>SC_9126_Tumor</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>SC_9126_TM_NB_SC_9126_Tumor_SC_9126_Normal</td>\n",
       "      <td>SC_9126</td>\n",
       "      <td>SC_9126_Tumor</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id         Center  NCBI_Build  Chromosome  \\\n",
       "0        A1BG               1  broad.mit.edu          37          19   \n",
       "1        A1BG               1  broad.mit.edu          37          19   \n",
       "2        A1BG               1  broad.mit.edu          37          19   \n",
       "3        A1BG               1  broad.mit.edu          37          19   \n",
       "4        A1CF           29974  broad.mit.edu          37          10   \n",
       "\n",
       "   Start_Position  End_position Strand Tumor_Sample_Barcode  \\\n",
       "0        58862934      58862934      +         TCGA-EJ-5499   \n",
       "1        58863660      58863660      +              MO_1012   \n",
       "2        58863782      58863782      +         TCGA-CH-5752   \n",
       "3        58864304      58864304      +          06-134H1_LN   \n",
       "4        52569681      52569681      +              SC_9126   \n",
       "\n",
       "         Matched_Norm_Sample_Barcode  ...       pos               type  \\\n",
       "0  PRAD-TCGA-EJ-5499-Normal-SM-1U3NZ  ...  58862934  Missense_Mutation   \n",
       "1                     MO_1012-Normal  ...  58863660  Missense_Mutation   \n",
       "2  PRAD-TCGA-CH-5752-Normal-SM-1U3IX  ...  58863782             Silent   \n",
       "3                    06-134A1_NORMAL  ...  58864304             Silent   \n",
       "4                     SC_9126_Normal  ...  52569681  Missense_Mutation   \n",
       "\n",
       "  classification ref_allele                            patient Primary_Met  \\\n",
       "0            SNP          G   PRAD-TCGA-EJ-5499-Tumor-SM-1U3IG     Primary   \n",
       "1            SNP          A  MO_1012-Tumor-Abdomen_wall_nodule  Metastasis   \n",
       "2            SNP          C   PRAD-TCGA-CH-5752-Tumor-SM-1U3ID     Primary   \n",
       "3            SNP          C                        06-134H1_LN  Metastasis   \n",
       "4            SNP          C                      SC_9126_Tumor  Metastasis   \n",
       "\n",
       "                                             pair_id      individual_id  \\\n",
       "0          PRAD-TCGA-EJ-5499-TP-NB-SM-1U3IG-SM-1U3NZ  PRAD-TCGA-EJ-5499   \n",
       "1  MO_1012_TM_NB_MO_1012-Tumor-Abdomen_wall_nodul...            MO_1012   \n",
       "2          PRAD-TCGA-CH-5752-TP-NB-SM-1U3ID-SM-1U3IX  PRAD-TCGA-CH-5752   \n",
       "3                        06-134H1_LN_06-134A1_NORMAL             06-134   \n",
       "4         SC_9126_TM_NB_SC_9126_Tumor_SC_9126_Normal            SC_9126   \n",
       "\n",
       "                         case_sample  Pair_Set_ID  \n",
       "0   PRAD-TCGA-EJ-5499-Tumor-SM-1U3IG  Prim_762017  \n",
       "1  MO_1012-Tumor-Abdomen_wall_nodule   Met_762017  \n",
       "2   PRAD-TCGA-CH-5752-Tumor-SM-1U3ID  Prim_762017  \n",
       "3                        06-134H1_LN   Met_762017  \n",
       "4                      SC_9126_Tumor   Met_762017  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(os.path.join(data_dir, filename), sep='\\t', low_memory=False, skiprows=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(95354, 56)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Hugo_Symbol',\n",
       " 'Entrez_Gene_Id',\n",
       " 'Center',\n",
       " 'NCBI_Build',\n",
       " 'Chromosome',\n",
       " 'Start_Position',\n",
       " 'End_position',\n",
       " 'Strand',\n",
       " 'Tumor_Sample_Barcode',\n",
       " 'Matched_Norm_Sample_Barcode',\n",
       " 'Variant_Type',\n",
       " 'Variant_Classification',\n",
       " 'Reference_Allele',\n",
       " 'Tumor_Seq_Allele1',\n",
       " 'Tumor_Seq_Allele2',\n",
       " 'Codon_Change',\n",
       " 'cDNA_Change',\n",
       " 'Protein_Change',\n",
       " 't_alt_count',\n",
       " 'allelicfraction',\n",
       " 't_total_count',\n",
       " 'newbase',\n",
       " 'effect_idx',\n",
       " 'newbase_idx',\n",
       " 'end',\n",
       " 'tum_allele1',\n",
       " 'tum_allele2',\n",
       " 'start',\n",
       " 'pat_idx',\n",
       " 'is_coding',\n",
       " 'is_flank',\n",
       " 'is_indel',\n",
       " 'is_ins',\n",
       " 'is_del',\n",
       " 'is_missense',\n",
       " 'is_nonsense',\n",
       " 'is_splice',\n",
       " 'is_silent',\n",
       " 'gene_idx',\n",
       " 'context_and_effect',\n",
       " 'context65',\n",
       " 'categ_idx',\n",
       " 'trackpos',\n",
       " 'categ',\n",
       " 'gene',\n",
       " 'chr',\n",
       " 'pos',\n",
       " 'type',\n",
       " 'classification',\n",
       " 'ref_allele',\n",
       " 'patient',\n",
       " 'Primary_Met',\n",
       " 'pair_id',\n",
       " 'individual_id',\n",
       " 'case_sample',\n",
       " 'Pair_Set_ID']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[col for col in df.columns]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hugo Symbol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The HUGO Gene Nomenclature Committee (HGNC) designates an official name and symbol (an abbreviation of the name) for each known human gene. The HGNC is a nonprofit organization funded by the U.S. National Human Genome Research Institute and the UK's Wellcome Trust. The Committee has named more than 19,000 of the estimated 20,000 to 25,000 protein-coding genes in the human genome.\n",
    "\n",
    "During the research process, genes often acquire several alternate names and symbols from researchers investigating the same gene. To resolve this confusion, the HGNC assigns a unique name and symbol to each human gene, which allows effective organization of genes in large databanks, aiding the advancement of research.\n",
    "\n",
    "Reference:\n",
    "\n",
    "&emsp;https://medlineplus.gov/genetics/understanding/mutationsanddisorders/naming/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46528    MAN2C1\n",
       "46548      MAOA\n",
       "81380     TECTA\n",
       "79150      STRC\n",
       "35382     HIP1R\n",
       "28471     FCGBP\n",
       "25120      EML5\n",
       "80542    TAS1R1\n",
       "67765    RASSF8\n",
       "90375     XIRP2\n",
       "Name: Hugo_Symbol, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Hugo_Symbol'].sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A1BG', 'A1CF', 'A2M', ..., 'ZYX', 'ZZEF1', 'ZZZ3'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Hugo_Symbol'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15741"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Hugo_Symbol'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Entrez Gene ID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Entrez Gene is the gene-specific database at the National Center for Biotechnology Information (NCBI), a division of the National Library of Medicine (NLM), located on the campus of the US National Institutes of Health (NIH) in Bethesda, MD, USA. Entrez Gene provides unique integer identifiers for genes and other loci (such as officially named mapped markers) for a subset of model organisms.\n",
    "\n",
    "Entrez Gene ID is an identifier for a gene per the NCBI Entrez database.\n",
    "\n",
    "The primary goals of Entrez Gene are to provide tracked, unique identifiers for genes of multiple genomes and to report information associated with those identifiers for unrestricted public use. The identifier that is assigned (GeneID) is an integer, and is species-specific. In other words, the integer assigned to dystrophin in human is different from that in any other species.\n",
    "\n",
    "Reference:\n",
    "\n",
    "&emsp;1. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1761442/\n",
    "2. https://www.wikidata.org/wiki/Property:P351"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70671    23429\n",
       "42959     3846\n",
       "50565    94025\n",
       "88864     8875\n",
       "7406     10396\n",
       "Name: Entrez_Gene_Id, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Entrez_Gene_Id'].sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chromosomes are a threadlike structure of nucleic acids and protein found in the nucleus of most living cells, carrying genetic information in the form of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([19, 10, 12, 22,  3,  1,  4, 15,  2, 17, 16,  6, 11,  7,  9, 23, 13,\n",
       "       14, 21, 20, 18,  5,  8, 24], dtype=int64)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>type</th>\n",
       "      <th>classification</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>patient</th>\n",
       "      <th>Primary_Met</th>\n",
       "      <th>pair_id</th>\n",
       "      <th>individual_id</th>\n",
       "      <th>case_sample</th>\n",
       "      <th>Pair_Set_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3675</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>266</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>6734123</td>\n",
       "      <td>6734123</td>\n",
       "      <td>+</td>\n",
       "      <td>PROS11496-6115321-SM-6CNQ5</td>\n",
       "      <td>PROS11496-6115321-Normal-SM-66WBA</td>\n",
       "      <td>...</td>\n",
       "      <td>6734123</td>\n",
       "      <td>Intron</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>PROS11496-6115321-Tumor-SM-6CNQ5</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>PROS11496-6115321-TM-NB-SM-6CNQ5-SM-66WBA</td>\n",
       "      <td>PROS11496-6115321</td>\n",
       "      <td>PROS11496-6115321-Tumor-SM-6CNQ5</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3676</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>266</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>6736812</td>\n",
       "      <td>6736812</td>\n",
       "      <td>+</td>\n",
       "      <td>SC_9076</td>\n",
       "      <td>SC_9076-NB</td>\n",
       "      <td>...</td>\n",
       "      <td>6736812</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>SC_9076-TM</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>SC_9076_TM_NB_SC_9076-TM_SC_9076-NB</td>\n",
       "      <td>SC_9076</td>\n",
       "      <td>SC_9076-TM</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20826</th>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>8653</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15023762</td>\n",
       "      <td>15023762</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGA-HC-8266</td>\n",
       "      <td>PRAD-TCGA-HC-8266-Normal-SM-2XTT7</td>\n",
       "      <td>...</td>\n",
       "      <td>15023762</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>PRAD-TCGA-HC-8266-Tumor-SM-2XTS9</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PRAD-TCGA-HC-8266-TP-NB-SM-2XTS9-SM-2XTT7</td>\n",
       "      <td>PRAD-TCGA-HC-8266</td>\n",
       "      <td>PRAD-TCGA-HC-8266-Tumor-SM-2XTS9</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20827</th>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>8653</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15023867</td>\n",
       "      <td>15023867</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGA-YL-A8HO</td>\n",
       "      <td>PRAD-TCGA-YL-A8HO-Normal-SM-5KLIH</td>\n",
       "      <td>...</td>\n",
       "      <td>15023867</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>PRAD-TCGA-YL-A8HO-Tumor-SM-5KMXW</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PRAD-TCGA-YL-A8HO-TP-NB-SM-5KMXW-SM-5KLIH</td>\n",
       "      <td>PRAD-TCGA-YL-A8HO</td>\n",
       "      <td>PRAD-TCGA-YL-A8HO-Tumor-SM-5KMXW</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20828</th>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>8653</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15028523</td>\n",
       "      <td>15028523</td>\n",
       "      <td>+</td>\n",
       "      <td>PR-3036</td>\n",
       "      <td>PR-3036-Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>15028523</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>PR-3036-Tumor</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PR-3036-TP-N</td>\n",
       "      <td>PR-3036</td>\n",
       "      <td>PR-3036-Tumor</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88535</th>\n",
       "      <td>UTY</td>\n",
       "      <td>7404</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15438342</td>\n",
       "      <td>15438343</td>\n",
       "      <td>+</td>\n",
       "      <td>05-214BB2_BONE</td>\n",
       "      <td>05-214A2_NORMAL</td>\n",
       "      <td>...</td>\n",
       "      <td>15438342</td>\n",
       "      <td>Intron</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GT</td>\n",
       "      <td>05-214BB2_BONE</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>05-214BB2_BONE_05-214A2_NORMAL</td>\n",
       "      <td>05-214</td>\n",
       "      <td>05-214BB2_BONE</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88536</th>\n",
       "      <td>UTY</td>\n",
       "      <td>7404</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15448190</td>\n",
       "      <td>15448190</td>\n",
       "      <td>+</td>\n",
       "      <td>SC_9172</td>\n",
       "      <td>SC_9172_Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>15448190</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>SC_9172_Tumor</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>SC_9172_TM_NB_SC_9172_Tumor_SC_9172_Normal</td>\n",
       "      <td>SC_9172</td>\n",
       "      <td>SC_9172_Tumor</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88537</th>\n",
       "      <td>UTY</td>\n",
       "      <td>7404</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>15478111</td>\n",
       "      <td>15478114</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGA-EJ-5507</td>\n",
       "      <td>PRAD-TCGA-EJ-5507-Normal-SM-1U3NW</td>\n",
       "      <td>...</td>\n",
       "      <td>15478111</td>\n",
       "      <td>Intron</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAAT</td>\n",
       "      <td>PRAD-TCGA-EJ-5507-Tumor-SM-1U3I5</td>\n",
       "      <td>Primary</td>\n",
       "      <td>PRAD-TCGA-EJ-5507-TP-NB-SM-1U3I5-SM-1U3NW</td>\n",
       "      <td>PRAD-TCGA-EJ-5507</td>\n",
       "      <td>PRAD-TCGA-EJ-5507-Tumor-SM-1U3I5</td>\n",
       "      <td>Prim_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91652</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>7544</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>2847155</td>\n",
       "      <td>2847155</td>\n",
       "      <td>+</td>\n",
       "      <td>SC_9144</td>\n",
       "      <td>SC_9144_Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>2847155</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>SC_9144_Tumor</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>SC_9144_TM_NB_SC_9144_Tumor_SC_9144_Normal</td>\n",
       "      <td>SC_9144</td>\n",
       "      <td>SC_9144_Tumor</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91653</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>7544</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>2847984</td>\n",
       "      <td>2847984</td>\n",
       "      <td>+</td>\n",
       "      <td>7951</td>\n",
       "      <td>PRAD-7951-Normal-SM-3SEWV</td>\n",
       "      <td>...</td>\n",
       "      <td>2847984</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>PRAD-7951-Tumor-SM-4B2BK</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>PRAD-7951-TR-NT-SM-4B2BK-SM-3SEWV</td>\n",
       "      <td>PRAD-7951</td>\n",
       "      <td>PRAD-7951-Tumor-SM-4B2BK</td>\n",
       "      <td>Met_762017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Hugo_Symbol  Entrez_Gene_Id         Center  NCBI_Build  Chromosome  \\\n",
       "3675        AMELY             266  broad.mit.edu          37          24   \n",
       "3676        AMELY             266  broad.mit.edu          37          24   \n",
       "20826       DDX3Y            8653  broad.mit.edu          37          24   \n",
       "20827       DDX3Y            8653  broad.mit.edu          37          24   \n",
       "20828       DDX3Y            8653  broad.mit.edu          37          24   \n",
       "...           ...             ...            ...         ...         ...   \n",
       "88535         UTY            7404  broad.mit.edu          37          24   \n",
       "88536         UTY            7404  broad.mit.edu          37          24   \n",
       "88537         UTY            7404  broad.mit.edu          37          24   \n",
       "91652         ZFY            7544  broad.mit.edu          37          24   \n",
       "91653         ZFY            7544  broad.mit.edu          37          24   \n",
       "\n",
       "       Start_Position  End_position Strand        Tumor_Sample_Barcode  \\\n",
       "3675          6734123       6734123      +  PROS11496-6115321-SM-6CNQ5   \n",
       "3676          6736812       6736812      +                     SC_9076   \n",
       "20826        15023762      15023762      +                TCGA-HC-8266   \n",
       "20827        15023867      15023867      +                TCGA-YL-A8HO   \n",
       "20828        15028523      15028523      +                     PR-3036   \n",
       "...               ...           ...    ...                         ...   \n",
       "88535        15438342      15438343      +              05-214BB2_BONE   \n",
       "88536        15448190      15448190      +                     SC_9172   \n",
       "88537        15478111      15478114      +                TCGA-EJ-5507   \n",
       "91652         2847155       2847155      +                     SC_9144   \n",
       "91653         2847984       2847984      +                        7951   \n",
       "\n",
       "             Matched_Norm_Sample_Barcode  ...       pos               type  \\\n",
       "3675   PROS11496-6115321-Normal-SM-66WBA  ...   6734123             Intron   \n",
       "3676                          SC_9076-NB  ...   6736812             Silent   \n",
       "20826  PRAD-TCGA-HC-8266-Normal-SM-2XTT7  ...  15023762  Missense_Mutation   \n",
       "20827  PRAD-TCGA-YL-A8HO-Normal-SM-5KLIH  ...  15023867  Nonsense_Mutation   \n",
       "20828                     PR-3036-Normal  ...  15028523  Missense_Mutation   \n",
       "...                                  ...  ...       ...                ...   \n",
       "88535                    05-214A2_NORMAL  ...  15438342             Intron   \n",
       "88536                     SC_9172_Normal  ...  15448190  Missense_Mutation   \n",
       "88537  PRAD-TCGA-EJ-5507-Normal-SM-1U3NW  ...  15478111             Intron   \n",
       "91652                     SC_9144_Normal  ...   2847155             Silent   \n",
       "91653          PRAD-7951-Normal-SM-3SEWV  ...   2847984    Frame_Shift_Del   \n",
       "\n",
       "      classification ref_allele                           patient Primary_Met  \\\n",
       "3675             DEL          A  PROS11496-6115321-Tumor-SM-6CNQ5  Metastasis   \n",
       "3676             SNP          G                        SC_9076-TM  Metastasis   \n",
       "20826            SNP          A  PRAD-TCGA-HC-8266-Tumor-SM-2XTS9     Primary   \n",
       "20827            SNP          G  PRAD-TCGA-YL-A8HO-Tumor-SM-5KMXW     Primary   \n",
       "20828            SNP          G                     PR-3036-Tumor     Primary   \n",
       "...              ...        ...                               ...         ...   \n",
       "88535            DEL         GT                    05-214BB2_BONE  Metastasis   \n",
       "88536            SNP          C                     SC_9172_Tumor  Metastasis   \n",
       "88537            DEL       AAAT  PRAD-TCGA-EJ-5507-Tumor-SM-1U3I5     Primary   \n",
       "91652            SNP          A                     SC_9144_Tumor  Metastasis   \n",
       "91653            DEL          G          PRAD-7951-Tumor-SM-4B2BK  Metastasis   \n",
       "\n",
       "                                          pair_id      individual_id  \\\n",
       "3675    PROS11496-6115321-TM-NB-SM-6CNQ5-SM-66WBA  PROS11496-6115321   \n",
       "3676          SC_9076_TM_NB_SC_9076-TM_SC_9076-NB            SC_9076   \n",
       "20826   PRAD-TCGA-HC-8266-TP-NB-SM-2XTS9-SM-2XTT7  PRAD-TCGA-HC-8266   \n",
       "20827   PRAD-TCGA-YL-A8HO-TP-NB-SM-5KMXW-SM-5KLIH  PRAD-TCGA-YL-A8HO   \n",
       "20828                                PR-3036-TP-N            PR-3036   \n",
       "...                                           ...                ...   \n",
       "88535              05-214BB2_BONE_05-214A2_NORMAL             05-214   \n",
       "88536  SC_9172_TM_NB_SC_9172_Tumor_SC_9172_Normal            SC_9172   \n",
       "88537   PRAD-TCGA-EJ-5507-TP-NB-SM-1U3I5-SM-1U3NW  PRAD-TCGA-EJ-5507   \n",
       "91652  SC_9144_TM_NB_SC_9144_Tumor_SC_9144_Normal            SC_9144   \n",
       "91653           PRAD-7951-TR-NT-SM-4B2BK-SM-3SEWV          PRAD-7951   \n",
       "\n",
       "                            case_sample  Pair_Set_ID  \n",
       "3675   PROS11496-6115321-Tumor-SM-6CNQ5   Met_762017  \n",
       "3676                         SC_9076-TM   Met_762017  \n",
       "20826  PRAD-TCGA-HC-8266-Tumor-SM-2XTS9  Prim_762017  \n",
       "20827  PRAD-TCGA-YL-A8HO-Tumor-SM-5KMXW  Prim_762017  \n",
       "20828                     PR-3036-Tumor  Prim_762017  \n",
       "...                                 ...          ...  \n",
       "88535                    05-214BB2_BONE   Met_762017  \n",
       "88536                     SC_9172_Tumor   Met_762017  \n",
       "88537  PRAD-TCGA-EJ-5507-Tumor-SM-1U3I5  Prim_762017  \n",
       "91652                     SC_9144_Tumor   Met_762017  \n",
       "91653          PRAD-7951-Tumor-SM-4B2BK   Met_762017  \n",
       "\n",
       "[61 rows x 56 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Chromosome']==24]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strand, Start Position, and End Position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An individual strand of DNA is referred to as positive-sense (also positive (+) or simply sense) if its nucleotide sequence corresponds directly to the sequence of an RNA transcript which is translated or translatable into a sequence of amino acids (provided that any thymine bases in the DNA sequence are replaced with uracil bases in the RNA sequence). The other strand of the double-stranded DNA molecule is referred to as negative-sense (also negative (−) or antisense), and is reverse complementary to both the positive-sense strand and the RNA transcript. It is actually the antisense strand that is used as the template from which RNA polymerases construct the RNA transcript, but the complementary base-pairing by which nucleic acid polymerization occurs means that the sequence of the RNA transcript will look identical to the sense strand, apart from the RNA transcript's use of uracil instead of thymine.\n",
    "\n",
    "The Sense strand is also known as coding strand, plus strand or non template strand, and is in the 5' --> 3' direction.\n",
    "\n",
    "The Antisense strand is also known as non coding strand, negative strand or template strand, and is in the 3' --> 5' direction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Start and End Positions denote the genomic coordinates of a gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For instance, to find the transcription start site (TSS) of a gene that spans the region of (in genomic coordinates) 10800 to 11200 in, say, chromosome 11, you should first consider the strand the gene is located in. Let's say the base at 10800 is A and the base at 11200 is C. If the gene is in the + strand, than 10800-11200 refer to the 5'-3' of the gene, hence the TSS is 10800 A. If the genes is in the - strand, then 10800-11200 (always for + strand) corresponds to the 3'-5' direction of the gene located in the complementary strand, hence the TSS is 11200 G (complementary to C)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15562</th>\n",
       "      <td>105150791</td>\n",
       "      <td>105150791</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13433</th>\n",
       "      <td>60781351</td>\n",
       "      <td>60781351</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42679</th>\n",
       "      <td>39538253</td>\n",
       "      <td>39538253</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74640</th>\n",
       "      <td>128694704</td>\n",
       "      <td>128694705</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11157</th>\n",
       "      <td>181732588</td>\n",
       "      <td>181732588</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5557</th>\n",
       "      <td>119112316</td>\n",
       "      <td>119112316</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7065</th>\n",
       "      <td>194126869</td>\n",
       "      <td>194126871</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75930</th>\n",
       "      <td>101631786</td>\n",
       "      <td>101631786</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Start_Position  End_position Strand\n",
       "15562       105150791     105150791      +\n",
       "13433        60781351      60781351      +\n",
       "42679        39538253      39538253      +\n",
       "74640       128694704     128694705      +\n",
       "11157       181732588     181732588      +\n",
       "5557        119112316     119112316      +\n",
       "7065        194126869     194126871      +\n",
       "75930       101631786     101631786      +"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Start_Position', 'End_position', 'Strand']].sample(8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A gene variant is a permanent change in the DNA sequence that makes up a gene. This type of genetic change used to be known as a gene mutation, but because changes in DNA do not always cause disease, it is thought that gene variant is a more accurate term. Variants can affect one or more DNA building blocks (nucleotides) in a gene.\n",
    "\n",
    "Type of Gene Variants:\n",
    "\n",
    "A. Classification based on cell type\n",
    "* `Germline variants` are variants in egg or sperm cells, and are able to be passed from parent to offspring.\n",
    "* `Somatic variants` are present only in specific cells and are generally not hereditary. Somatic variants are acquired at some point during a person’s lifetime, often due to environmental factors or errors in cell division. Some somatic variants will lead to the development of diseases such as cancer. Both germline and somatic variants have been implicated in many cancers. \n",
    "\n",
    "B. Classification based on type of alteration\n",
    "* `Single nucleotide polymorphisms (SNPs)` or `Single base-pair substitution` — this is a substitution of a single nucleotide for another. It can be `transition` (interchange of the purine (Adenine/Guanine) or pyrimidine (Cytosine/Thymine) nucleic acids) or `transversion` (interchange of a purine and pyrimidine nucleic acid).\n",
    "* `Double nucleotide polymorphisms (DNPs)`\n",
    "* `Triple nucleotide polymorphisms (TNPs)`\n",
    "* `Insertion` — this changes the DNA sequence by adding one or more nucleotides to the gene. As a result, the protein made from the gene may not function properly.\n",
    "* `Deletion` — this changes the DNA sequence by removing at least one nucleotide in a gene. Small deletions remove one or a few nucleotides within a gene, while larger deletions can remove an entire gene or several neighboring genes. The deleted DNA may alter the function of the affected protein or proteins.\n",
    "* `Deletion-Insertion` — this variant occurs when a deletion and insertion happen at the same time in the same location in the gene. In a deletion-insertion variant, at least one nucleotide is removed and at least one nucleotide is inserted. However, the change must be complex enough to differ from a simple substitution. The resulting protein may not function properly. A deletion-insertion (delins) variant may also be known as an insertion-deletion (indel) variant.\n",
    "* `Frameshift` — if the length of an indel is not a multiple of three, then it will likely result in a frameshift. A frameshift changes the reading frame of all the remaining bases resulting in the rest of the gene being incorrectly translated. This will drastically alter the product of translation, often leading to targeted decay of the mRNA .\n",
    "* `Copy number variation` — this is a duplication or deletion that changes the number of copies of a particular DNA segment within the genome. Structural variations have been implicated in a number of conditions, including polycystic kidney disease, cardiomyopathies, amyotrophic lateral sclerosis (ALS) and some cases of intellectual disability.\n",
    "* `Duplication` — this occurs when a stretch of one or more nucleotides in a gene is copied and repeated next to the original DNA sequence. This type of variant may alter the function of the protein made from the gene.\n",
    "* `Repeat expansion` — some regions of DNA contain short sequences of nucleotides that are repeated a number of times in a row. For example, a trinucleotide repeat is made up of sequences of three nucleotides, and a tetranucleotide repeat is made up of sequences of four nucleotides. A repeat expansion is a variant that increases the number of times that the short DNA sequence is repeated. This type of variant can cause the resulting protein to function improperly.\n",
    "\n",
    "\n",
    "References:\n",
    "* https://medlineplus.gov/genetics/understanding/mutationsanddisorders/possiblemutations/\n",
    "* https://www.ebi.ac.uk/training/online/courses/human-genetic-variation-introduction/what-is-genetic-variation/types-of-genetic-variation/\n",
    "* https://www.garvan.org.au/research/kinghorn-centre-for-clinical-genomics/learn-about-genomics/for-gp/genetics-refresher-1/types-of-variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SNP', 'INS', 'DEL', 'DNP', 'TNP'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Variant_Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Missense_Mutation', 'Silent', 'Splice_Site', 'Intron',\n",
       "       'Frame_Shift_Del', 'Nonsense_Mutation', \"3'UTR\", \"5'UTR\",\n",
       "       'In_Frame_Del', 'Frame_Shift_Ins', 'In_Frame_Ins', 'RNA',\n",
       "       'De_novo_Start_InFrame', 'Stop_Codon_Ins', 'Nonstop_Mutation',\n",
       "       'Start_Codon_Del', 'Start_Codon_Ins', 'Stop_Codon_Del',\n",
       "       'De_novo_Start_OutOfFrame', 'lincRNA'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Variant_Classification'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* `Missense mutation` — this is a point mutation in which a single nucleotide change results in a codon that codes for a different amino acid. It is when the change of a single base pair causes the substitution of a different amino acid in the resulting protein. This amino acid substitution may have no effect, or it may render the protein nonfunctional.\n",
    "* `Silent mutation` — this occurs when the change of a single DNA nucleotide within a protein-coding portion of a gene does not affect the sequence of amino acids that make up the gene's protein.\n",
    "* `Splice-site mutation` — this is a genetic alteration in the DNA sequence that occurs at the boundary of an exon and an intron (splice site). This change can disrupt RNA splicing resulting in the loss of exons or the inclusion of introns and an altered protein-coding sequence. Also called splice-site variation.\n",
    "* `Intron mutation` — this is a mutation to an intron. Generally there should not be any effect as the introns are spliced out.\n",
    "* `Frame shift deletion` — basically a deletion that occurs in a frameshift-wise.\n",
    "* `Nonsense mutation` — this is a change in DNA that causes a protein to terminate or end its translation earlier than expected. This is a common form of mutation in humans and in other animals that causes a shortened or nonfunctional protein to be expressed. It is also known as `stop mutation`.\n",
    "* `3' UTR mutation` — also known as `three prime untranslated region mutation`. The 3'-untranslated region (3'-UTR) of mRNA contains regulatory elements that are essential for the appropriate expression of many genes. These regulatory elements are involved in the control of nuclear transport, polyadenylation status, subcellular targetting as well as rates of translation and degradation of mRNA. A mutation to this region can cause diseases.\n",
    "* `5' UTR mutation` — also known as `five prime untranslated region mutation`. Like the 3' UTR mutation, the 5' UTR mutation can cause genetic defects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7 (tensorflow)",
   "language": "python",
   "name": "tensorflow"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
